src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,225 @@
+
+// Dump format:
+// Nodes
+// "node_fields":
+//   [ "type", "name", "id", "self_size", "edge_count", "trace_node_id", "detachedness" ]
+
+struct Node {
+    string type;
+    string name;
+    size_t id;
+    size_t self_size;
+    int edge_count;
+    size_t trace_node_id;
+    // whether the node is attached or dettached from the main application state
+    // TODO: .... meaning not yet understood.
+    // https://github.com/nodejs/node/blob/5fd7a72e1c4fbaf37d3723c4c81dce35c149dc84/deps/v8/include/v8-profiler.h#L739-L745
+    int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+};
+
+// Edges
+// "edge_fields":
+//   [ "type", "name_or_index", "to_node" ]
+
+struct Edge {
+    string type;
+    size_t name_or_index; // essentially 'from'
+    size_t to_node;
+};
+
+//template<typename K, typename V>
+//auto find_or_insert_iter(unordered_map<K,V>& map, const K &key) {
+//}
+
+struct StringTable {
+    typedef unordered_map<string, size_t> MapType;
+
+    MapType map;
+    vector<string> strings;
+
+    StringTable() {}
+    StringTable(std::initializer_list<string> strs) : strings(strs) {
+        for (const auto& str : strs) {
+            map.insert({str, map.size()});
+        }
+    }
+
+    size_t find_or_create_string_id(string key) {
+        auto val = map.find(key);
+        if (val == map.end()) {
+            val = map.insert(val, {key, map.size()});
+            strings.push_back(key);
+        }
+        return val->second;
+    }
+
+    void print_json_array(JL_STREAM *stream, bool newlines) {
+        jl_printf(stream, "[");
+        bool first = true;
+        for (const auto &str : strings) {
+            if (first) {
+                first = false;
+            } else {
+                jl_printf(stream, ",");
+                if (newlines) {
+                    jl_printf(stream, "\n");
+                }
+            }
+            jl_printf(stream, "\"%s\"", str.c_str());
+        }
+        jl_printf(stream, "]");
+    }
+};
+
+struct HeapSnapshot {
+public:
+
+// private:
+    vector<Node> nodes;
+
+    vector<Edge> edges;
+
+    StringTable names;
+    StringTable node_types = {"object"};
+    StringTable edge_types = {"property"};
+    unordered_set<size_t> seen_node_ids;
+};
+
+
+// TODO: Do we need to refer to nodes by their index in the node array?
+//size_t find_or_create_node_id(HeapSnapshot& snapshot, string key) {
+//    return find_or_insert_iter(snapshot.nodes_map, key)->second;
+//}
+
+HeapSnapshot *g_snapshot = nullptr;
+
+JL_DLLEXPORT void jl_gc_take_heap_snapshot(JL_STREAM *stream) {
+    // Enable snapshotting
+    g_snapshot = new HeapSnapshot();
+
+    // Do GC, which will callback into record_edge_to_gc_snapshot()...
+    jl_gc_collect(JL_GC_FULL);
+
+    // When we return, the snapshot is full
+    // Dump the snapshot
+    serialize_heap_snapshot(stream, *g_snapshot);
+    
+    // Disable snapshotting
+    g_snapshot = nullptr;
```

Comment:
does this actually free the snapshot (and everything inside it?) just want to make sure we're not leaking all that